[Team4/BE] 숙소 조회 목업 컨트롤러 및 테스트

Dae-Hwa opened the pull request:
안녕하세요 프레디입니다. 

숙소 조회 목업 소스와 컨트롤러를 작성했습니다.

pr을 작은 단위로 했는데, 이전의 미션들처럼 특정한 단위가 없다보니, '이것만 하고 해야지..' '저것만 더 하고...' 하며 점점 pr이 늦어지게 되고 막바지에는 구현에 급급해졌던 것 같아 이번에는 작은 단위로 시작해보려고 합니다.

목업만 있는데 굳이 pr을 보내는 또 하나의 이유는 테스트에 대한 피드백을 받아보고 싶어서인데요, 컨트롤러부터 단위 테스트로 시작하여 Mock 객체를 만들어 나가고 싶었습니다. 그런데 이런 식의 테스트를 처음 작성해보다보니 어려움도 많았고 잘 하고 있는것인지 모르겠어서요..

현재는 서비스를 Mock 객체로 만들어 컨트롤러를 고립시킨 상태에서 테스트를 진행했고, assertion은 리스폰스 바디를 확인 하는 것으로 했습니다. 이 과정에서 mockMvc는 사용하기 까다로운 부분이 있어서 잭슨으로 객체로 매핑한 뒤 검사했습니다.

컨트롤러 단위 테스트를 진행할 때 확인해볼 수 있는 부분들이
- 올바르게 매개변수가 가공되어 서비스까지 전달되는지
- 올바른 컨텐츠 타입과 응답이 오는지
- 리스폰스 바디가 제대로 담기는지

이 정도로 생각되어 해당 부분들에 대해서 테스트를 진행하고 있습니다.

Dae-Hwa committed:
chore: .gitignore 생성

Dae-Hwa committed:
chore: 파트별 폴더 생성

Dae-Hwa committed:
chore: notion 링크 추가

Dae-Hwa committed:
chore: 프로젝트 기본 세팅

Dae-Hwa/airbnb/#3

Dae-Hwa committed:
chore: heroku 설정 파일 추가

Dae-Hwa/airbnb/#4

Dae-Hwa committed:
chore: 헤로쿠 배포 action 스크립트 추가

Dae-Hwa/airbnb/#7

Dae-Hwa committed:
fix: mockup api 깃헙 액션 스크립트 수정

Dae-Hwa committed:
fix: 불필요한 컨트롤러 제거

Dae-Hwa committed:
feat: 숙소검색 추가

협의 후 변경 예정

DaeHwa/airbnb/#9

Dae-Hwa committed:
chore: swagger 세팅 추가

Dae-Hwa committed:
chore: 커밋 템플릿 추가

Dae-Hwa committed:
Merge branch 'team-4' of https://github.com/Dae-Hwa/airbnb into BE/dev

Dae-Hwa committed:
chore: README.md에 커밋 템플릿 관련 내용 추가

Dae-Hwa committed:
docs: README에 팀원 및 팀 문서 링크 추가

Dae-Hwa committed:
chore: 커밋 템플릿에 타입 설명 추가

Dae-Hwa committed:
docs: README.md에 커밋 템플릿 예시 사진 업데이트

Dae-Hwa committed:
Merge branch 'team-4' of https://github.com/Dae-Hwa/airbnb into BE/dev

Dae-Hwa committed:
refactor: DummyDataFactory추가 및 enum 타입 추가

숙소 타입과 화장실 타입 enum으로 변경.

Dae-Hwa/airbnb/#9

Dae-Hwa committed:
refactor: 소스 정리

controller, domain으로 패키지 이동.

Dae-Hwa/airbnb/#9

Dae-Hwa committed:
feat: AccommodationService 추가 및 DummyData 수정

더미데이터에서 id 미리 넣도록 수정. 테스트 고려함.

Dae-Hwa/airbnb/#9

Dae-Hwa committed:
test: AccommodationConrollerTest 추가

MockMvc 사용. jackson 사용하여 결과 객체로 파싱하여 비교. 비교 위해
DTO에 equals 추가.
MockMvc사용 시 필터 지정하지 않으면 인코딩이 UTF8로 지정되지 않는데,
필터 지정을 위해서는 appicationContext를 넣어줘야 하여 WebMvcTest로
진행이 불가능. 때문에 application.properties에 강제 인코딩 옵션 추가.
인코딩 타입 판단 위해 Const에 상수로 UTF8 붙은 인코딩 타입 추가하여
비교.

Dae-Hwa/airbnb/#9

Dae-Hwa committed:
feat: Accommodation readAll에 매개변수 추가, DTO 추가 및 검증

하이버네이트 Validator 사용하여 검증. 검증결과 핸들러 추가 및 여기서 사용할 ErrorResponse 객체 생성함. mockMvc로 assertion이 힘든 부분이 있어 Jackson ObjectMapper 사용.

Dae-Hwa committed:
Merge pull request #28 from Dae-Hwa/BE/mockup/main

[BE] 숙소 전체 조회 목업

wheejuni commented on BE/airbnb-be/src/main/java/com/codesquad/airbnb/accommodation/domain/AccommodationOption.java:
```diff
@@ -0,0 +1,151 @@
+    public AccommodationOption() {
+    }
+
+    public AccommodationOption(int capacity, int pricePerNight, AccommodationType accommodationType, int bedroomCount, int restroomCount, RestroomType restroomType, boolean hasKitchen, boolean hasInternet, boolean hasAirconditioner, boolean hasHairdrier) {
```
어우 엄청나네요. 이 생성자를 가릴 수 있는 방법이 있으면 좋겠습니다.

Dae-Hwa replied:
빌더로만 생성할 수 있도록 protected로 변경해줬습니다.

추가적으로 들었던 생각은 현재 DTO내부에서 이걸 사용하고 있는데, 이런 식으로 작성하면 레이어가 섞여서 안 좋은 구조인가요?
그리고 근본적으로 DTO에 관해서도 생각이 들었는데, DTO를 리스폰스나 리퀘스트 데이터 매핑 용도로 사용하게된다면 이런 용도의 DTO는 굳이 도메인이나 엔티티 객체와 1:1로 생성하지 않아도 될 것 같다는 생각이 들었습니다.
DB와 객체가 미스매칭되는 것 처럼 API와 도메인 객체들이 완전히 부합하지 않을 수 있기 때문이라는 생각이 들었는데, 혹시 브라이언의 의견은 어떠신지 여쭙고 싶습니다!

wheejuni commented on BE/airbnb-be/src/main/java/com/codesquad/airbnb/common/Consts.java:
```diff
@@ -0,0 +1,7 @@
+import org.springframework.http.MediaType;
+
+public class Consts {
+    public static final String CONTENT_TYPE_JSON_UTF8 = MediaType.APPLICATION_JSON_VALUE + ";charset=UTF-8";
```
이미 있네요.
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/MediaType.html#APPLICATION_JSON_UTF8

그런데 주석을 보니 최신의 브라우저들에서는 `charset=UTF-8` 자체가 필요 없다고 합니다. 클라이언트 개발자들과 토론해 보시면 좋겠습니다.

Dae-Hwa replied:
맞습니다. deprecated됐는데, 억지로 사용하려고 했습니다...
이유는 MockMvc의 결과가 인코딩이 제대로 되지 않았기 때문이었는데, 더 찾아보니 deprecated된 것도 그렇고 RFC 8259 표준 관련 문제인 것 같습니다.
현재는 MockMvc를 사용하지 않기때문에 없애줬는데, 이 부분은 좀 더 찾아보겠습니다

wheejuni commented on BE/airbnb-be/src/test/java/com/codesquad/airbnb/accommodation/controller/AccommodationControllerTest.java:
```diff
@@ -0,0 +1,151 @@
+    @MockBean
+    private AccommodationService accommodationService;
+
+    private ObjectMapper objectMapper = new ObjectMapper();
```
프로젝트 전반에서 메세지 컨버터에 주입되어 사용될 `ObjectMapper` 를 그대로 사용하는 것이,
실제 실행 환경과 가장 유사한 테스트 환경을 만드는데에 도움이 될 것 같습니다.

`ObjectMapper` 도 주입받을 수 있습니다. 주입받기를 시도해주시고 안된다면 _어떻게 하면 되게할 수 있을지_ 고민해주세요.

Dae-Hwa replied:
기본적으로 Jackson 설정이 되어 있으니 당연히 주입 받아서 사용할 수 있을 것인데 이 부분 생각을 전혀 못하고 있었네요.
만약 안 된다면 설정을 따로 해주며 Bean 등록을 해주면 될 것 같은데 이 부분 말씀해주신게 맞을까요?

그리고 드물겠지만 혹시나 테스트 이외의 프로덕션 코드에서 사용할 일이 있을 때, 별도의 설정이 필요한 경우가 아니라면 빈으로 등록된 것을 재사용하는게 여러모로 좋을 것 같은데, 제가 생각한게 맞는지 여쭤보고 싶습니다.

wheejuni commented on BE/airbnb-be/src/test/java/com/codesquad/airbnb/accommodation/controller/AccommodationControllerTest.java:
```diff
@@ -0,0 +1,151 @@
+    @ParameterizedTest
+    @MethodSource("readAllProvider")
+    void readAll(String url, AccommodationRequestDTO accommodationRequestDTO, List<AccommodationResponseDTO> valueWillReturn, List<AccommodationResponseDTO> expected) throws Exception {
+        BDDMockito.given(accommodationService.readAll(accommodationRequestDTO))
+                .willReturn(valueWillReturn);
```
테스트를 통해 이루고 싶은 목표가 무엇인지요? 테스트 커버리지의 향상? 스프링 프레임워크 학습? 
이것을 체크해보고 싶은 이유는 컨트롤러에 대한 테스트는 _통합_ 테스트, _인수_ 테스트로의 목적을 지니는 경우가 많은데 지금 소스 코드를 보니 다른 도메인 객체나 컴포넌트에 대한 유닛 테스트가 모두 생략되어 있고 컨트롤러 테스트부터 작성하셨다는 느낌이 들기 때문입니다. 

많은 경우에 컨트롤러에 대한 테스트는 **블랙박스 테스트** 입니다. 즉, 컨트롤러 테스트 자체가 어떠한 전제조건이나 상태를 갖지 않고 의도한 요청에 대해 의도한 응답이 오는지 평가한 후, 프로젝트의 완성도를 짐작할 수 있는 테스트인 것입니다. 
쉽게 말해 특정 엔드포인트에 대한 컨트롤러 테스트가 성공했다면 이는 단순히 컨트롤러 로직의 성공을 의미하는 것이 아니라 요청이 프로젝트 전반의 한 싸이클을 돌면서... 그에 관계된 모든 컴포넌트들이 잘 작동하고 있고 잘 통합되어 하나의 소프트웨어로 출고될 준비를 마쳤다는 의미가 됩니다. 

그런데 지금 작성하신 컨트롤러 테스트를 보면 핵심 로직에 대해서는 `Mockito` 를 활용하여 모두 바이패스하였고 실제 프로젝트가 상용에 들어갈 준비가 됐음을 평가할 수 있는 어떠한 장치도 없어 보입니다(컨트롤러 테스트를 통한 인수 테스트를 빡세게 하시는 분들은 실제 데이터 등록이나 적재를 위해 사용할 API를 테스트 코드에 호출해 테스트 컨텍스트에 데이터를 쌓아두고 테스트를 시작합니다). 

만약 어떤 이유에서건 테스트 커버리지를 비약적으로 끌어올리고 테스트를 좀 더 꼼꼼히 작성하기로 결심하셨다면 그 출발점은 **도메인 객체** 와 **도메인 컴포넌트** 여야 의미있지 컨트롤러를 고립시켜 컨트롤러에 대한 테스트를 먼저 작성하시는 것에 대해 큰 의미를 부여하기는 어렵겠다는 생각이 듭니다.

Dae-Hwa replied:
저도 단위테스트 작성하며 좀 의아한 부분이 있었는데, 그 부분은 의견 주신 것 보고나서 어느정도 이해가 된 것 같습니다. 감사합니다!
말씀하신대로 엔드포인터 테스트로써 컨트롤러에 특별한 로직이 들어가는게 아니라면 단위테스트가 필요없다고 생각되어 통합테스트로 다 바꿔줬습니다.
처음에 작성하며 기대했던 부분 중에 도메인 영역과 컨트롤러 영역을 떨어뜨릴 수 있었으면 좋겠다는 것이 있었는데 이렇게하면 그 부분도 어느정도 만족이 되는 것 같습니다. 현재 로직 상으로는 앞으로 서비스는 지금 정해둔 api에 부합하는 DTO 리턴만 하면 되니까요. 물론 여기에 서비스 단위 테스트도 더해져야 할 것 같긴 합니다. 그 때는 레포지토리를 Mock으로 만들어서 사용하면 될 것 같다고 생각했습니다.

인수테스트 도구로 TestRestTemplate을 사용했습니다. RestAssured가 좀 더 쓰기 편해보이고 많이 쓰이는 것 같긴한데, 스프링에 내장 된 도구를 사용해보고 싶었습니다. 이전에 OAuth 호출하며 RestTemplate을 사용해봤는데, 좀 더 만져보면 좋겠다는 생각도 있었습니다.

그리고 인수테스트로 바뀌면서 서비스에 로직이 조금 들어가게 됐는데, 이 부분은 도메인을 만들면서 수정해나가도록 하겠습니다.

wheejuni reviewed (changes requested):
새로운 시도를 해보신듯한데 대단히 수고 많으셨습니다. 👍 
아무래도 현재 진척 단계를 보아하니 도메인 로직에 대한 구현은 아직 이뤄지지 않은 것 같고 mock 응답/로직을 기반으로 하여 컨트롤러에 대해서만 우선 테스트를 시도해보신 것 같습니다. 
나쁘지 않은데 이에 대한 저의 생각은 주절주절 인라인 코멘트로 남겨 보았습니다. 

테스트는 유닛 테스트 -> 통합 테스트 -> 인수 테스트 의 순서를 거치는 것이 원칙인데요, 뭐 원칙이라고 말은 했지만....
실제로 테스트를 많이 해 보시면 이 플로우로 가야 코드 작성자 본인도 정신의 편안함이 온다는 사실을 알게 되실 겁니다. 

컨트롤러에 대한 테스트는 많은 경우에 _인수 테스트_ 에 해당합니다. 
이를 기반으로 댓글 남겨두었으니 확인 부탁드리겠습니다. 

아... 생각해보니 지금 우리는 mock 기반으로 클라이언트 개발자와 협의하여 개발을 진행하고 있는데 어떻게 도메인 로직을 작성하고 테스트를 하느냐... 라는 질문을 하실 수 있겠어요.

**다 방법이 있습니다.** 어떤 방법이 있을까요?

_키워드_

- Spring Profile
- `@ConditionalOnProperty`
- `@Profile`

Dae-Hwa committed:
feat: 컨트롤러 인수테스트로 변경 및 서비스 수정

컨트롤러는 테스트는 인수테스트로 진행하는게 더 적절하다고 한다. https://github.com/codesquad-members-2021/airbnb/pull/28#discussion_r638778521
인수테스트에는 RestAssured와 TestRestTemplate를 사용할 수 있다고 하는데,
TestRestTemplate은 의존성 추가 없이 사용할 수 있어 선택

Dae-Hwa/airbnb/#35

Dae-Hwa committed:
fix: 불필요한 import 제거

Dae-Hwa committed:
feat: 불필요한 클래스 삭제

Dae-Hwa/airbnb/#35

Dae-Hwa committed:
feat: AccommodationOption 정리

불필요한 setter 제거 및 생성자 접근자 protected로 변경. 빌더가 외부에
있어 protected로 변경함.

Dae-Hwa/airbnb/#35

Dae-Hwa committed:
feat: 컨텐츠 타입 charset 제거

Dae-Hwa/airbnb/#35

Dae-Hwa committed:
Merge branch 'BE/mockup/accommocation' into BE/mockup/main

Dae-Hwa committed:
fix: TestRestTemplate에 로케일 추가

헤로쿠는 미국서버라 로케일이 en_us로 지정되어있음. 따라서 클라이언트에서
미 지정시 영어 메세지 리턴하기때문에 요청 시 로케일 추가해줌. aws배포도
동일할 것이라 생각하여 이렇게 처리함.

Dae-Hwa/airbnb/#35

Dae-Hwa committed:
Merge pull request #36 from Dae-Hwa/BE/mockup/main

[BE] 숙소 조회 리뷰 반영

Dae-Hwa commented:
배포 관련해서 서버와 로컬의 locale 에 관해서 기초적인 부분을 놓치고 있었습니다.
이것 때문에 좀 오래걸렸습니다 ㅠㅠ

말씀해주신 부분 중 스프링 프로필을 사용해서 도메인과 mockup을 분리하라고 하신 부분이 이해가 잘 되지 않아 다시 질문드립니다.

목업과 실제 도메인의 개발이 이뤄지는 부분의 프로필을 다르게 잡고가면 서로 다른 테스트를 짤 수 있고 그러면 배포시에도 용이하니 그렇게 하는게 더 좋다는 의미인가요?

Dae-Hwa committed:
feat: 목업과 도메인 개발 환경 분리
